### ISSUE TITLE
If a user is a member of multiple groups in the IdP, the user property is comma separated and we default to "all users"

### ISSUE THREAD

**Issue by paoliniluis** [metabase/metabase#40517]
### Describe the bug

When a user belong to more than 1 group, then the user property of group membership is a comma separated value (with a "/" also in Keycloak, don't know how it works in other systems). As we seem to compare against exact matches, then the group mapping doesn't work

### To Reproduce

1) set up the keycloak demo
2) create 2 groups in keycloak
3) add a user to 2 groups
4) set up group mapping in the server and then authenticate with the user
5) see that the user was not mapped to any group, but to all users

### Expected behavior

We should add the user to the combination of all the groups that the user is on

### Logs

NA

### Information about your Metabase installation

```JSON
- v49 but pretty sure it has always been like that
```

### Severity

P2, as only 1 user has reported this

### Additional context

NOTE: AFTER WE SHIP THIS WE SHOULD PUT A NOTE IN THE RELEASE NOTES ABOUT THIS, AS I'M AFRAID THAT CUSTOMERS WITH SSO WILL BE IMPACTED OF THE CHANGE

**Comment by luizarakaki** [comment]
We spent a long time trying to reproduce this
[Slack thread](https://metaboat.slack.com/archives/C013N8XL286/p1711532641515569)

Unless there are better reproduction steps, we can't work on it

**Comment by jonmauney** [comment]
👋 hello @luizarakaki!  I'm the person that reported this.  Thanks for looking into it. Are you able to access the Zendesk ticket we have with @Tony-Metabase on the matter? (looks like ticket ID: 25860.) We've got some details about the config in there- it looks to us like we've followed the docs, and the `metabaseGroups` attr on the SAML assertion that is being sent to Metabase appears to be generating a comma-separated list of groups as expected. Finally, we believe we have group mappings set up correctly on the Metabase SAML config side of things.

What we're seeing is that single group coming from Okta is **correctly** assigning the user to the group, but multiple groups end up **unassigning all groups** (other than All Users, of course), leading us to believe that something is malformed / being processed incorrectly.  I certainly won't rule out it's something on our side that is malformed, but it looks correct from what we can see re: [these](https://www.metabase.com/docs/latest/people-and-groups/saml-okta#example-of-mapping-multiple-groups-to-metabase) docs). I looked through logs and it does not appear the SAML assertions are appearing in there; is there any way we can see what Metabase is receiving to validate that the metabaseGroups attr is structured per the documentation?

**Comment by luizarakaki** [comment]
My best hypothesis now is that this is Okta-specific.

We tried to reproduce this with Keycloak and it worked as expected (multiple group assignment).
Maybe Okta formats the XML differently from Keycloak.

Does this SAML assertion contain sensitive data? Can you send it to us? (can send through zendesk)

**Comment by jonmauney** [comment]
Thanks @luizarakaki.  The group names are certainly not sensitive; not sure about the rest of the content, but will examine / post in a redacted form if necessary. I'll work with my IT team today to see if we can get a dump of the XML for direct testing / examination (I am unfortunately, but appropriately, not an Okta admin in our org!) In the meantime, here are some details that may help (or perhaps uncover obvious mistakes): 

The metabaseGroups attr is being generated using  Okta Expression Language (it's not manually input):

<img width="445" alt="image" src="https://github.com/metabase/metabase/assets/4520575/3d9118d4-610a-48db-8834-63e66ffea52f">

The full metabaseGroups attr (for this user) looks like this:

```
Metabase - Admins,Metabase - Owners,Metabase - All Users
```

...and is generated using Okta Expression Language (group IDs redacted and list is shortened):
```
getFilteredGroups({"groupID1", "groupID2", "groupID3", "groupID4", "groupID5"}, "[group.name]", 100)
```

The groupIDs are resolving to the list above.

Note that I thought there could have been an issue with symbol conversion with spaces, but renaming the groups to remove spaces did not exhibit different behavior.

**Comment by luizarakaki** [comment]
Right... Yeah, my first guess would be special characters causing problems. We had issues in the past with some special characters...

We are working internally to try to reproduce this using Okta dev tools. You'll be back to you on your ticket

**Comment by jonmauney** [comment]
@luizarakaki just got a response back from our IT team, it does look like the shape of the attr coming back from Okta differs from what is expected.

Okta is producing a comma separated string as a single-element attribute: 

```xml
        <saml2:Attribute Name="metabaseGroups" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:basic">
            <saml2:AttributeValue
                xmlns:xs="http://www.w3.org/2001/XMLSchema"
                xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:type="xs:string">Metabase - Admins,MetabaseOwners
            </saml2:AttributeValue>
        </saml2:Attribute>
```

In the Metabase[ test](https://github.com/metabase/metabase/blob/80e690534a18ffb666124208e215d68b42182bce/test_resources/saml-test-response-new-user-with-groups.xml#L91) for multiple group assignment, it looks like the attr is an array:
```xml
<saml:Attribute Name="GroupMembership" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:basic">
        <saml:AttributeValue xsi:type="xs:anyType">group_1</saml:AttributeValue>
        <saml:AttributeValue xsi:type="xs:anyType">group_2</saml:AttributeValue>
      </saml:Attribute>
```

I'm going to see if there is a different data type we can assign in Okta that will unpack that list into an array rather than a comma-separated string.

**Comment by luizarakaki** [comment]
Thanks, this is useful. I'll ping the team.
Let me know if you can find a way to change this structure in Okta

**Comment by luizarakaki** [comment]
Is there are reason to add `[ ]` around group.name?
getFilteredGroups({"groupID1", "groupID2", "groupID3", "groupID4", "groupID5"}, "[group.name]", 100)

In the docs it says it returns an array, but it doesn't add these []
https://developer.okta.com/docs/reference/okta-expression-language/#group-functions

**Comment by Tony-metabase** [comment]
Hey @jonmauney,

Actually it does work. I am not sure how you are configuring users ... but let me share with you what i did. First i create a user and assign departments to it (in my case `Metabase-Admin,Metabase-Test`:

<img width="1505" alt="image" src="https://github.com/metabase/metabase/assets/110378427/d39b6897-a14b-4cf7-9e90-209765d79062">

there are also firstName, lastName and email but those you mentioned you already setup but just for consistency these are part of the User Profile:

<img width="1512" alt="image" src="https://github.com/metabase/metabase/assets/110378427/75bc0d0a-9d3c-403a-9204-2a60c482640a">


I then proceed to create the application which is that SAML 2.0 that we will connect to Metabase and setup the following attributes the following:

<img width="1507" alt="image" src="https://github.com/metabase/metabase/assets/110378427/9ffe8946-615d-485f-bccb-1872828b150e">

Note that department is the User profile attribute in which i am defining the groups. I need to perform an `Arrays.flatten` else metabase thinks its a string and will not understand multiple groups.

i then setup authentication with SAML and defining the group mappings:

<img width="1255" alt="image" src="https://github.com/metabase/metabase/assets/110378427/492e52c1-2080-44b6-b2ce-56816f5b9457">

Then when i login with this user i can see it being added to both groups:

<img width="1512" alt="image" src="https://github.com/metabase/metabase/assets/110378427/342e475e-b5ff-4d0a-b252-c0f0710d7142">

So this seems to be a false alarm or maybe the way we are creating users/groups is different?

**Comment by jonmauney** [comment]
Thanks both @luizarakaki @Tony-metabase! Will get with our IT team today to try this; will get back to you.

**Comment by jonmauney** [comment]
Unfortunately I can't get time with IT to test until Thursday afternoon.  One thing I found; it's possible there is an undocumented feature flag that is not enabled in our Okta account called `SAML_SUPPORT_ARRAY_ATTRIBUTES`.

Similar Okta support tickets / third-party how-to guides [here](https://support.okta.com/help/s/question/0D54z00007K7QA4CAN/saml-assertion-memberof-format-issue?language=en_US) and [here](https://support.okta.com/help/s/question/0D51Y00009sknKkSAI/same-variable-name-with-multiple-tags?language=en_US), and recognition in third-party guides [here](https://support.cloudinary.com/hc/en-us/articles/4405194731026-SAML-Provisioning-with-Okta-Tips-for-setting-user-attributes#:~:text=SAML_SUPPORT_ARRAY_ATTRIBUTES) and [here](https://github.com/rapid7/awsaml/blob/master/README.md).  We'll assess that alongside trying your suggestions on Thursday.

**Comment by jonmauney** [comment]
Agreed that it should be the default @Tony-metabase; only thing is that our Okta account is many years old and I could see them not turning on the flag for older accounts if it could change behavior of existing Apps.  Will report back when we sit down to dig in!

**Comment by Tony-metabase** [comment]
No worries! I will close this issue for now since it seems that it's working but lets keep the thread open on our ticket :) and we can troubleshoot from there!

We can open another one which would be more specific for your usecase if that's not the issue

**Comment by bc-rp** [comment]
Hi,

We had the same issue and have resolved it as follows:

1. Have Okta enable the feature flag SAML_SUPPORT_ARRAY_ATTRIBUTES for your tenant
2. Configure the group attribute statement
- Name: MetabaseGroupName
- Name Format: Basic
- Value: Arrays.flatten(getFilteredGroups({"Grp_1", "Grp_2", "Grp_3"}, "group.name", 100))

With the above configuration we are now able to have users assigned to multiple groups.
